### DISCUSSION TITLE
Implementing scene queries using immediate mode

### DISCUSSION THREAD

**Discussion by zopsicle** [NVIDIA-Omniverse/PhysX#146] [Q&A] [answered]
When using immediate mode, PxBroadPhase can accelerate body pair generation, but its API does not provide AABB tree traversal. Is there a structure that can be used to accelerate queries against many geometries when using immediate mode? A dynamic AABB tree that is agnostic of leaf type, similar to PxBroadPhase, would be sufficient. When using PxScene, one can choose a data structure using PxPruningStructureType, but I can’t find a public API to use those data structures directly.

**Comment by PierreTerdiman** [comment, marked as answer]
_**When using immediate mode, PxBroadPhase can accelerate body pair generation, but its API does not provide AABB tree traversal.**_

The PxBroadPhase API does not provide AABB tree traversal because the various broadphase implementations of this API do not necessarily use an AABB tree. Related documentation: https://nvidia-omniverse.github.io/PhysX/physx/5.1.3/docs/SceneQueries.html#rationale-for-decoupled-broadphases-scene-query-system

_**Is there a structure that can be used to accelerate queries against many geometries when using immediate mode? A dynamic AABB tree that is agnostic of leaf type, similar to PxBroadPhase, would be sufficient.**_

You can use either a standalone PxBVH (simple option) or Gu::QuerySystem (more complex option). Related documentation:

https://nvidia-omniverse.github.io/PhysX/physx/5.1.3/docs/SceneQueries.html#standalone-pxbvh

https://nvidia-omniverse.github.io/PhysX/physx/5.1.3/docs/SceneQueries.html#low-level-query-system

There are various snippets mentioned in the above links that should get you started.
